File: merkle-trie/src/main/java/net/consensys/cava/trie/NodeVisitor.java
package net.consensys.cava.trie;

import net.consensys.cava.bytes.Bytes;
import net.consensys.cava.concurrent.AsyncResult;

interface NodeVisitor<V> {

  AsyncResult<Node<V>> visit(ExtensionNode<V> extensionNode, Bytes path);

  AsyncResult<Node<V>> visit(BranchNode<V> branchNode, Bytes path);

  AsyncResult<Node<V>> visit(LeafNode<V> leafNode, Bytes path);

  AsyncResult<Node<V>> visit(NullNode<V> nullNode, Bytes path);
}

File: merkle-trie/src/main/kotlin/net/consensys/cava/trie/experimental/NodeVisitor.kt
package net.consensys.cava.trie.experimental

import net.consensys.cava.bytes.Bytes

internal interface NodeVisitor<V> {

  suspend fun visit(extensionNode: ExtensionNode<V>, path: Bytes): Node<V>

  suspend fun visit(branchNode: BranchNode<V>, path: Bytes): Node<V>

  suspend fun visit(leafNode: LeafNode<V>, path: Bytes): Node<V>

  suspend fun visit(nullNode: NullNode<V>, path: Bytes): Node<V>
}
